js/plugins/mcp/src/client/index.ts

```diff
@@ -0,0 +1,242 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SSEClientTransportOptions } from '@modelcontextprotocol/sdk/client/sse.js';
+import { StdioServerParameters } from '@modelcontextprotocol/sdk/client/stdio.js';
+import { Transport } from '@modelcontextprotocol/sdk/shared/transport.js';
+import { Genkit, ToolAction } from 'genkit';
+import { logger } from 'genkit/logging';
+import { GenkitMcpClient, McpClientOptions } from './client';
+export type { SSEClientTransportOptions, StdioServerParameters, Transport };
+
+export interface McpClientManagerOptions {
+  /**
+   * An optional name for this client manager. This name is primarily for
+   * logging and identification purposes within Genkit.
+   * Defaults to 'genkitx-mcp'.
+   */
+  name?: string;
+  /**
+   * A record where each key is a client-side name for an MCP server,
+   * and each value is an `McpClientOptions` object defining how to connect
+   * to that server and how it should behave.
+   * This allows the manager to handle connections to multiple, uniquely named MCP servers.
+   */
+  mcpClients?: Record<string, McpClientOptions>;
+}
+
+/**
+ * Manages connections to multiple MCP (Model Context Protocol) servers.
+ * Each server connection is individually configured and managed by an instance of `GenkitMcpClient`.
+ * This manager provides a centralized way to initialize, update, and interact with these clients.
+ *
+ * It allows for dynamic registration of tools from all connected and enabled MCP servers
+ * into a Genkit instance.
+ */
+export class GenkitMcpClientManager {
+  name: string;
+  private _clients: Record<string, GenkitMcpClient> = {};
+  private _readyListeners: {
+    resolve: () => void;
+    reject: (err: Error) => void;
+  }[] = [];
+  private _ready = false;
+
+  constructor(options: McpClientManagerOptions) {
+    this.name = options.name || 'genkitx-mcp';
+
+    if (options.mcpClients) this.updateClients(options.mcpClients);
+  }
+
+  /**
+   * Returns a Promise that resolves when the manager has attempted to connect
+   * to all configured clients, or rejects if a critical error occurs during
+   * the initial connection phase.
+   */
+  async ready() {
+    if (this._ready) return;
+    return new Promise<void>((resolve, reject) => {
+      this._readyListeners.push({ resolve, reject });
+    });
+  }
+
+  /**
+   * Connects to a single MCP server defined by the provided configuration.
+   * If a server with the same name already exists, it will be disconnected first.
+   * Stores the client and transport references internally. Handles connection errors
+   * by marking the server as disabled.
+   * @param serverName The name to assign to this server connection.
+   * @param config The configuration object for the server.
+   */
+  async connectClient(serverName: string, config: McpClientOptions) {
```

Comment:
From a nomenclature perspective I think we should consider ourselves the client and the thing the dev is trying to interact with the server.

So you're connecting a server, not a client -- the client is an implementation detail.

But in terms of names I'd just keep it simple with `connect(serverName: string, config: {...})` and do the same throughout.

Reply:
Thanks, I was a bit confused myself with where I was landing with "client" vs "server" terminology. Updated.